### PR TITLE
Use optimized queries in link endpoints

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Very similar to #1275 but for links

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203007126736610/1203225514907869/f) _(internal)_

## 🚫 Blocked by

- #1258 
- #1255
- #1267
- #1268
- #1269
- #1271
- #1272 
- #1273 
- #1228 
- #1274
- #1275

## 🔍 What does this change?

Applies the same changes as #1275 but for links instead

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 1]
```diff
@@ -1,42 +1,70 @@
...
+            .query_raw(&statement, parameters.iter().copied())
+            .await
+            .into_report()
+            .change_context(QueryError)?
+            .map(|row| row.into_report().change_context(QueryError))
+            .and_then(|row| async move {
+                let link_type_id = VersionedUri::from_str(row.get(0))
+                    .into_report()
+                    .change_context(QueryError)?;
+                let source_entity_id = row.get(1);
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:57]: Are the orders of the results always stable? As in do they always follow the order of `add_selection_path`?

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 2]
```diff
@@ -1,42 +1,70 @@
...
-                } else {
-                    bail!(
-                        Report::new(ExpressionError)
-                            .attach_printable("does not result in a boolean value")
-                            .change_context(QueryError)
-                    );
-                }
+        let mut compiler = SelectCompiler::new();
+        compiler.add_selection_path(
+            &LinkQueryPath::Type(LinkTypeQueryPath::VersionedUri),
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:30]: Is this selection is required to get distinctness through the primary key?

**Review by thehabbos007 — APPROVED** [review]
Thanks, Tim. I have some questions, but looks good to merge :)

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:57]: Yes, they are stable 🙂

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:30]: Yes, it's required. I avoided adding extra methods like `primary_keys` or something like that.

**Pull request merged**